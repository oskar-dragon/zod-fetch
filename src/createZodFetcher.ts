import { Schema, z, ZodObject, ZodRawShape, ZodTypeAny } from 'zod';

export type AnyFetcher = (...args: any[]) => any;

export type DefaultFetcher = (
  ...args: Parameters<typeof fetch>
) => Promise<Response>;

export const defaultFetcher: DefaultFetcher = async (...args) => {
  const response = await fetch(...args);

  if (!response.ok) {
    throw new Error(`Request failed with status ${response.status}`);
  }

  return response.json();
};

export type ZodFetcher<TFetcher extends AnyFetcher> = <TData>(
  schema: Schema<TData>,
  ...args: Parameters<TFetcher>
) => Promise<TData>;

export function createZodFetcher<TFetcher extends AnyFetcher>(
  fetcher: AnyFetcher = defaultFetcher,
): ZodFetcher<TFetcher> {
  return async (schema, ...args) => {
    const data = await fetcher(...args);
    return schema.parse(data);
  };
}
