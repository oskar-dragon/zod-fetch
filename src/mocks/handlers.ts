import { http, HttpResponse } from 'msw';

const baseURL = 'https://example.com';

export const getHelloPath = baseURL + '/user';
export const getHello = http.get(getHelloPath, () => {
  return HttpResponse.json({ name: 'John' });
});

export const getHelloErrorPath = getHelloPath + '/error';
export const getHelloError = http.get(getHelloErrorPath, () => {
  return new HttpResponse(null, {
    status: 404,
    statusText: 'User not found',
  });
});

export default [getHello];
