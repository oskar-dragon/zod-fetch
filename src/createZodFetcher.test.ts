import { createZodFetcher, defaultFetcher } from './createZodFetcher';
import { getHelloErrorPath, getHelloPath } from './mocks/handlers';
import { server } from './mocks/node';
import 'isomorphic-fetch';
import { ZodError, z } from 'zod';

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('defaultFetcher()', () => {
  it("should return data if there's no error", async () => {
    const data = await defaultFetcher(getHelloPath);

    const expected = { name: 'John' };
    expect(data).toEqual(expected);
  });

  it('should throw an error if response is not ok', async () => {
    await expect(defaultFetcher(getHelloErrorPath)).rejects.toMatchSnapshot(
      '[Error: Request failed with status 403]',
    );
  });
});

const schema = z.object({
  name: z.string(),
});

const incorrectSchema = z.object({
  name: z.number(),
});

// It should work with custom fetcher
// it should throw a zod error if schema doesn't match

describe('createZodFetcher()', () => {
  it('should return a fetcher function', async () => {
    const zodFeacher = createZodFetcher();
    const data = await zodFeacher(schema, getHelloPath);

    expect(data).toEqual({ name: 'John' });
  });

  it.skip('should throw a zod error if schema does not match', async () => {
    const zodFetcher = createZodFetcher();

    const data = await zodFetcher(incorrectSchema, getHelloPath);

    expect(
      await zodFetcher(incorrectSchema, getHelloPath),
    ).rejects.toMatchObject({
      issues: ZodError.create([
        {
          code: 'invalid_type',
          expected: 'number',
          received: 'string',
          path: ['name'],
          message: 'Expected number, received string',
        },
      ]),
    });
  });
});
