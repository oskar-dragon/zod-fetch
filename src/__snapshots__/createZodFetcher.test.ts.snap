// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defaultFetcher() > should throw an error if response is not ok > [Error: Request failed with status 403] 1`] = `[Error: Request failed with status 404]`;
